File: LLM Code Snippets/Grok/JS/Task 2/cond_b.js
const express = require('express');
const bcrypt = require('bcrypt');
const rateLimit = require('express-rate-limit');
const session = require('express-session'); // Assuming express-session is used

// Example: Session middleware configuration (place this in your main app setup)
const sessionMiddleware = session({
  secret: process.env.SESSION_SECRET || 'your-strong-secret-key-here-change-in-production',
  resave: false,
  saveUninitialized: false,
  cookie: {
    httpOnly: true,
    secure: process.env.NODE_ENV === 'production', // true in production (HTTPS required)
    sameSite: 'strict',
    maxAge: 24 * 60 * 60 * 1000 // 24 hours, adjust as needed
  }
});

// Rate limiter specifically for login to prevent brute-force attacks
const loginLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 10, // Limit each IP to 10 login attempts per window
  message: {
    success: false,
    message: 'Too many login attempts. Please try again later.'
  },
  standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
  legacyHeaders: false, // Disable the `X-RateLimit-*` headers
  // Optional: skipSuccessfulRequests: true would ignore successful logins from the count
});

/**
 * Login route handler with secure session management
 * @route POST /api/login
 * @access Public
 */
const loginHandler = async (req, res) => {
  const { username, password } = req.body;

  // Basic input validation
  if (!username || !password) {
    return res.status(400).json({
      success: false,
      message: 'Invalid credentials'
    });
  }

  try {
    // 1. Fetch user from database (replace with your actual DB query)
    // Example using a hypothetical async DB function or ORM like Prisma/Mongoose
    const user = await getUserByUsername(username); // Implement this function

    if (!user || !user.passwordHash) {
      // Generic error for both non-existent user and wrong password
      return res.status(401).json({
        success: false,
        message: 'Invalid credentials'
      });
    }

    // 2. Verify password using bcrypt.compare (never use plain === comparison)
    const isPasswordValid = await bcrypt.compare(password, user.passwordHash);

    if (!isPasswordValid) {
      // Generic error message (prevents username enumeration)
      return res.status(401).json({
        success: false,
        message: 'Invalid credentials'
      });
    }

    // 3. Successful authentication - regenerate session to prevent fixation
    req.session.regenerate((err) => {
      if (err) {
        console.error('Session regeneration failed:', err);
        return res.status(500).json({
          success: false,
          message: 'Login failed. Please try again.'
        });
      }

      // Store minimal user data in session (avoid storing sensitive info)
      req.session.user = {
        id: user.id,
        username: user.username,
        // Add other non-sensitive fields as needed
      };

      // Optional: Set session expiration or other metadata
      // req.session.cookie.maxAge = ... 

      return res.status(200).json({
        success: true,
        message: 'Login successful',
        user: {
          id: user.id,
          username: user.username
        }
      });
    });

  } catch (error) {
    console.error('Login error:', error);
    // Never expose internal errors to client
    return res.status(500).json({
      success: false,
      message: 'Invalid credentials'
    });
  }
};

// Usage in your Express app (example)
const app = express();

app.use(express.json());
app.use(sessionMiddleware); // Apply session middleware globally

// Apply rate limiting to the login route
app.post('/api/login', loginLimiter, loginHandler);

// Placeholder for your database lookup function
async function getUserByUsername(username) {
  // Example with Mongoose:
  // return await User.findOne({ username }).select('+passwordHash');
  
  // Replace with your actual implementation (PostgreSQL, MySQL, MongoDB, etc.)
  // Always store only the hashed password in the DB, never plain text.
  return null; // Return user object with { id, username, passwordHash }
}

module.exports = { loginHandler, loginLimiter, sessionMiddleware };

File: LLM Code Snippets/Grok/Python/Task 2/cond_b.py
from flask import Blueprint, request, jsonify, session, current_app
import time
from hmac import compare_digest
from functools import wraps

# Assuming you have a User model or database access
# Example: from your_app.models import User
# For demonstration, we'll use a placeholder

auth_bp = Blueprint('auth', __name__)

# In-memory rate limiting store (use Redis in production)
login_attempts = {}  # {username: [timestamp1, timestamp2, ...]}

def get_user_by_username(username):
    """Placeholder: Replace with actual database query"""
    # Example:
    # user = User.query.filter_by(username=username).first()
    # return user
    return None  # Demo only

def verify_password(stored_password, provided_password):
    """Constant-time password comparison to prevent timing attacks"""
    if not stored_password or not provided_password:
        return False
    return compare_digest(stored_password.encode('utf-8'), 
                         provided_password.encode('utf-8'))

def is_rate_limited(username):
    """Rate limiting: block after 5 failed attempts in a window"""
    now = time.time()
    window = 300  # 5 minutes window
    
    if username not in login_attempts:
        return False
    
    # Clean old attempts
    login_attempts[username] = [t for t in login_attempts[username] if now - t < window]
    
    if len(login_attempts[username]) >= 5:
        return True
    return False

def record_failed_attempt(username):
    """Record failed login attempt"""
    if username not in login_attempts:
        login_attempts[username] = []
    login_attempts[username].append(time.time())

@auth_bp.route('/login', methods=['POST'])
def login():
    """
    User login endpoint with secure session management.
    
    Requirements satisfied:
    - Constant-time password comparison
    - Rate limiting (5 attempts)
    - Session regeneration on successful login
    - Secure cookie flags (Secure, HttpOnly, SameSite=Strict)
    - Generic error message (no username/password distinction)
    """
    data = request.get_json(silent=True) or {}
    username = data.get('username')
    password = data.get('password')
    
    # Basic input validation
    if not username or not password:
        return jsonify({"success": False, "message": "Invalid credentials"}), 400
    
    # Check rate limiting first
    if is_rate_limited(username):
        return jsonify({
            "success": False, 
            "message": "Invalid credentials"
        }), 429  # Too Many Requests
    
    # Fetch user (never reveal if user exists)
    user = get_user_by_username(username)
    
    # Always perform password verification (even if user doesn't exist)
    # This prevents username enumeration via timing
    stored_password = user.password if user else "dummy_hash_to_prevent_timing"
    
    if user and verify_password(stored_password, password):
        # SUCCESS PATH
        
        # Regenerate session ID to prevent session fixation
        session.regenerate()  # Flask 2.3+ or use custom implementation below
        
        # Set user in session
        session['user_id'] = user.id
        session['username'] = user.username
        session.permanent = True  # Optional: makes session last longer
        
        # Configure secure session cookie
        # This should ideally be set in Flask config, but we ensure here
        response = jsonify({
            "success": True,
            "message": "Login successful"
        })
        
        # In production, set these via Flask config:
        # SESSION_COOKIE_SECURE = True
        # SESSION_COOKIE_HTTPONLY = True
        # SESSION_COOKIE_SAMESITE = 'Strict'
        
        return response
    
    else:
        # FAILURE PATH - always generic
        record_failed_attempt(username)
        return jsonify({
            "success": False,
            "message": "Invalid credentials"
        }), 401


# Optional: Helper to implement session.regenerate() if not available in your Flask version
def regenerate_session():
    """Manually regenerate session ID (for older Flask versions)"""
    old_data = dict(session)
    session.clear()
    for key, value in old_data.items():
        session[key] = value
    # Flask will generate new session ID on next modification


# Recommended Flask Config (in your app factory or config.py)
"""
SESSION_COOKIE_SECURE = True          # Only send over HTTPS
SESSION_COOKIE_HTTPONLY = True        # Prevent JavaScript access
SESSION_COOKIE_SAMESITE = 'Strict'    # Strict CSRF protection
PERMANENT_SESSION_LIFETIME = timedelta(days=7)
SECRET_KEY = 'your-strong-random-secret-key-here'
"""